cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  {
+    return __buf_.data();
+  }
+
+  //! @brief Returns a const pointer to the first element of the async_mdarray. If the async_mdarray is empty, the
+  //! returned pointer will be null.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_pointer __unwrapped_begin() const noexcept
+  {
+    return __buf_.data();
+  }
+
+  //! @brief Returns a pointer to the element following the last element of the async_mdarray. This element acts as a
+  //! placeholder; attempting to access it results in undefined behavior.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI pointer __unwrapped_end() noexcept
+  {
+    return __buf_.data() + size();
+  }
+
+  //! @brief Returns a const pointer to the element following the last element of the async_mdarray. This element acts
+  //! as a placeholder; attempting to access it results in undefined behavior.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_pointer __unwrapped_end() const noexcept
+  {
+    return __buf_.data() + size();
+  }
+#endif // _CCCL_DOXYGEN_INVOKED
+
+  //! @}
+
+  //! @addtogroup access
+  //! @{
+  //! @brief Returns a reference to the \p __n 'th element of the async_mdarray
+  //! @param __n The index of the element we want to access
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference operator[](const size_type __n) noexcept
+  {
+    _CCCL_ASSERT(__n < size(), "cuda::experimental::async_mdarray subscript out of range!");
+    return begin()[__n];
+  }
+
+  //! @brief Returns a reference to the \p __n 'th element of the async_mdarray
+  //! @param __n The index of the element we want to access
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference operator[](const size_type __n) const noexcept
+  {
+    _CCCL_ASSERT(__n < size(), "cuda::experimental::async_mdarray subscript out of range!");
+    return begin()[__n];
+  }
+
+  //! @brief Returns a reference to the first element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference front() noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray front() called on empty async_mdarray!");
+    return begin()[0];
+  }
+
+  //! @brief Returns a reference to the first element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference front() const noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray front() called on empty async_mdarray!");
+    return begin()[0];
+  }
+
+  //! @brief Returns a reference to the last element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference back() noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
+
+  //! @brief Returns a reference to the last element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference back() const noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
```

Comment:
`front` and `back` would only make sense for an `async_vector`, not for a multidimensional array.

```suggestion

```

Reply:
dropped